profile header buttons reorder

shohelbuddyboss opened the pull request:
### All Submissions:

* [x] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [x] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Fixes #1351 .

### Proof Screenshots or Video

https://www.loom.com/share/e09420f42818418597b0dfdd5ff32a33

<!-- Mark completed items with an [x] -->

### Changelog entry

> Added Profile header buttons reorder

ankitguptaindia commented:
@sourabhmatolia Please assign any dev to code review of this PR.

So we can start QA then.

mehulkaklotar commented on src/bp-templates/bp-nouveau/includes/xprofile/functions.php:
```diff
@@ -43,3 +43,55 @@ function bp_nouveau_xprofile_enqueue_scripts() {
+ * @return mixed|void
+ */
+
+function bb_profile_header_buttons() {
+
+	$buttons = apply_filters( 'bb_profile_header_buttons', array(
+		'member_friendship' => __( 'Connect', 'buddyboss' ),
+		'member_follow'     => __( 'Follow', 'buddyboss' ),
+		'private_message'   => __( 'Message', 'buddyboss' ),
+		'member_switch'     => __( 'View As', 'buddyboss' ),
+	) );
+
+	return $buttons;
```
@shohelbuddyboss We should not get static buttons array here. Is there any way we can have dynamic buttons for user?

The function name always starts with `bp_` so here it should be `bp_get_profile_header_buttons`

shohelbuddyboss replied:
@mehulkaklotar I tried to use existing function to get the button, but those buttons not appear at the customizer as buttons appear at the profile header based on various condition. So it's the best way, used filter to add remove buttons here.

Added the prefix bp_ instead of bb_ (bb means buddyboss)

mehulkaklotar commented on src/bp-templates/bp-nouveau/includes/xprofile/functions.php:
```diff
@@ -43,3 +43,55 @@ function bp_nouveau_xprofile_enqueue_scripts() {
+ * @return mixed|void
+ */
+
+function bb_profile_header_buttons() {
```
@shohelbuddyboss this functions should be in `bp-templates/bp-nouveau/includes/customizer.php`

shohelbuddyboss replied:
moved and renamed, new function name `bp_get_profile_header_buttons();`

mehulkaklotar commented on src/bp-templates/bp-nouveau/includes/xprofile/functions.php:
```diff
@@ -43,3 +43,55 @@ function bp_nouveau_xprofile_enqueue_scripts() {
+
+add_action( 'wp_ajax_save_profile_header_buttons_order', 'save_profile_header_buttons_order' );
+
+function save_profile_header_buttons_order(){
```
@shohelbuddyboss this functions should be in `bp-templates/bp-nouveau/includes/customizer.php`

Again, The function name always starts with bp_

shohelbuddyboss replied:
If move this function to `bp-templates/bp-nouveau/includes/customizer.php` the function is not working as expected, it's through undefined function from this file. so moved it to `bp-templates/bp-nouveau/includes/functions.php`

Added bp_

mehulkaklotar commented on src/bp-templates/bp-nouveau/includes/xprofile/functions.php:
```diff
@@ -43,3 +43,55 @@ function bp_nouveau_xprofile_enqueue_scripts() {
+	wp_send_json_success();
+}
+
+function bb_get_profile_header_buttons(){
```
@shohelbuddyboss this functions should be in bp-templates/bp-nouveau/includes/customizer.php

Again, The function name always starts with bp_

shohelbuddyboss replied:
If move this function to bp-templates/bp-nouveau/includes/customizer.php the function is not working as expected, it's through undefined function from this file. so moved it to bp-templates/bp-nouveau/includes/functions.php

Added bp_

new function name `bp_get_profile_header_buttons_by_order()`

mehulkaklotar commented on src/bp-templates/bp-nouveau/js/customizer.js:
```diff
@@ -93,6 +93,44 @@ window.wp = window.wp || {};
+			}
+		} ).disableSelection();
+
+		$( document ).on( 'change', '#bp_profile_header_order', function() {
+			var buttons = $( this ).val();
+
+
+			$.post( ajaxurl, { buttons : buttons, action : 'save_profile_header_buttons_order' }, function(){
+
+			} );
+
+		});
```
@shohelbuddyboss the controls in customizer are listen by the WP customizer js lib. can you check if we don't do this without checking with actual wp customizer. There should be a hidden or named field to store values for the order and it will sync automatically with customizer.

shohelbuddyboss replied:
@mehulkaklotar 

This is now deprecated as the changes are now syncing with customizer automatically.

shohelbuddyboss commented:
Hi @mehulkaklotar 

Sync profile header buttons reorder when sort from the customizer, manual ajax is no more required.
added bp_ prefix instead of bb
moved some function to `bp-templates/bp-nouveau/includes/functions.php`

Please review again. Now this is better then previous one.

mehulkaklotar commented on src/bp-templates/bp-nouveau/includes/customizer.php:
```diff
@@ -83,13 +91,30 @@ function bp_nouveau_customize_register( WP_Customize_Manager $wp_customize ) {
 			'transport'         => 'refresh',
 			'type'              => 'option',
 		),
+		'bp_nouveau_appearance[profile_header_buttons]' => array(
+			'index'             => 'profile_header_buttons',
+			'capability'        => 'bp_moderate',
+			'sanitize_callback' => 'bp_nouveau_sanitize_nav_order',
+			'transport'         => 'refresh',
+			'type'              => 'option',
+		),
```
```suggestion
		'bp_nouveau_appearance[user_profile_actions_display]' => array(
			'index'             => 'user_profile_actions_display',
			'capability'        => 'bp_moderate',
			'sanitize_callback' => 'bp_nouveau_sanitize_nav_order',
			'transport'         => 'refresh',
			'type'              => 'option',
		),
```

shohelbuddyboss replied:
Changed as per suggestion

mehulkaklotar commented on src/bp-templates/bp-nouveau/includes/customizer.php:
```diff
@@ -215,6 +240,15 @@ function bp_nouveau_customize_register( WP_Customize_Manager $wp_customize ) {
 			'type'     => 'user',
 		),
+
+		'profile_header_buttons'   => array(
+			'class'    => 'BP_Nouveau_Profile_Header_Customize_Control',
+			'label'    => __( 'Reorder the profile header buttons.', 'buddyboss' ),
+			'section'  => 'bp_nouveau_user_profile_header',
+			'settings' => 'bp_nouveau_appearance[profile_header_buttons]',
+			'type'     => 'header_button',
+		),
```
```suggestion
		'user_profile_actions_display'   => array(
			'class'    => 'BP_Nouveau_Profile_Header_Customize_Control',
			'label'    => __( 'Reorder the profile actions.', 'buddyboss' ),
			'section'  => 'bp_nouveau_user_profile_header',
			'settings' => 'bp_nouveau_appearance[user_profile_actions_display]',
			'type'     => 'header_button',
		),
```

shohelbuddyboss replied:
Changed as per suggestion

mehulkaklotar commented on src/bp-templates/bp-nouveau/includes/customizer.php:
```diff
@@ -270,3 +305,24 @@ function bp_nouveau_customizer_enqueue_scripts() {
+ * @return mixed|void
+ */
+
+function bp_get_profile_header_buttons() {
```
```suggestion
function bp_nouveau_customizer_user_profile_actions() {
```

shohelbuddyboss replied:
Changed function name

mehulkaklotar commented on src/bp-templates/bp-nouveau/includes/functions.php:
```diff
@@ -570,6 +570,8 @@ function bp_nouveau_get_appearance_settings( $option = '' ) {
 		'user_nav_display'   => 0, // O is default (horizontally). 1 is vertically.
 		'user_nav_order'     => array(),
 		'user_nav_hide'      => array(),
+		'profile_header_buttons'      => array(),
+		'profile_header_order'      => '',
```
```suggestion
		'user_profile_actions_display'      => array(),
		'user_profile_actions_order'      => '',
```

shohelbuddyboss replied:
Changed as per suggestion

mehulkaklotar commented on src/bp-templates/bp-nouveau/includes/functions.php:
```diff
@@ -1537,3 +1539,25 @@ function bp_nouveau_set_nav_item_order( $nav = null, $order = array(), $parent_s
+
+function bp_get_profile_header_buttons_by_order() {
+	$bp_nouveau_appearance     = bp_get_option( 'bp_nouveau_appearance' );
+	$profile_header_btn_orders = isset( $bp_nouveau_appearance['profile_header_order'] )
+		? $bp_nouveau_appearance['profile_header_order'] : '';
```
```suggestion
	$profile_header_btn_orders = isset( $bp_nouveau_appearance['user_profile_actions_order'] )
		? $bp_nouveau_appearance['user_profile_actions_order'] : '';
```

shohelbuddyboss replied:
Changed as per suggestion

mehulkaklotar commented on src/bp-templates/bp-nouveau/includes/functions.php:
```diff
@@ -1537,3 +1539,25 @@ function bp_nouveau_set_nav_item_order( $nav = null, $order = array(), $parent_s
+ * @return mixed|void
+ */
+
+function bp_get_profile_header_buttons_by_order() {
```
```suggestion
function bp_nouveau_get_user_profile_actions() {
```

shohelbuddyboss replied:
Changed as per suggestion

mehulkaklotar reviewed (changes requested):
@shohel please do suggested changes here and make neccessary adjustment here and there like function names changes at all places.

shohelbuddyboss commented on src/bp-templates/bp-nouveau/includes/customizer.php:
```diff
@@ -42,6 +44,12 @@ function bp_nouveau_customize_register( WP_Customize_Manager $wp_customize ) {
+			'title'       => __( 'Profile Header', 'buddyboss' ),
+			'panel'       => 'bp_nouveau_panel',
+			'priority'    => 50,
+			'description' => __( 'Re-order the profile header menu.', 'buddyboss' ),
```
Description changed

shohelbuddyboss commented on src/bp-templates/bp-nouveau/includes/customizer.php:
```diff
@@ -270,3 +305,24 @@ function bp_nouveau_customizer_enqueue_scripts() {
+		'member_follow'     => __( 'Follow', 'buddyboss' ),
+		'private_message'   => __( 'Message', 'buddyboss' ),
+		'member_switch'     => __( 'View As', 'buddyboss' ),
+	) );
```
Component checked

shohelbuddyboss commented on src/bp-templates/bp-nouveau/includes/customizer.php:
```diff
@@ -83,13 +91,30 @@ function bp_nouveau_customize_register( WP_Customize_Manager $wp_customize ) {
+			//'sanitize_callback' => 'bp_nouveau_sanitize_nav_order',
+			'transport'         => 'refresh',
+			'type'              => 'option',
+		),
```
Sanitized added. I would add later.

shohelbuddyboss commented:
Hi @mehulkaklotar 

I've made the necessary changes, please review the changes.

mehulkaklotar commented on src/bp-templates/bp-nouveau/includes/profile-header-customizer-controls.php:
```diff
@@ -0,0 +1,85 @@
+/**
+ * Customizer controls
+ *
+ * @since BuddyPress 3.0.0
```
@shohel change the since tags and BuddyBoss

shohelbuddyboss replied:
@mehulkaklotar  Done

shohelbuddyboss commented on src/bp-templates/bp-nouveau/includes/customizer.php:
```diff
@@ -270,3 +302,41 @@ function bp_nouveau_customizer_enqueue_scripts() {
+	//Member switch button
+	$buttons['member_switch'] = __( 'View As', 'buddyboss' );
+
+	$buttons = apply_filters( 'bp_get_user_profile_actions_display', $buttons );
```
@mehulkaklotar Done

mehulkaklotar reviewed (dismissed):
@shohel This looks good! 🚀 

In my testing, it looks the way it suppose to be.

@ankitguptaindia we need testing here!

ankitguptaindia commented:
@shohel 

1.  When using this branch then having below warnings-  https://i.imgur.com/4Ai2nhL.jpg
`Warning: array_search() expects parameter 2 to be array, string given in /Users/ank/Local Sites/julytest/app/public/wp-content/plugins/buddyboss-platform/src/bp-templates/bp-nouveau/includes/members/template-tags.php on line 144`
2. When trying to re-order profiles tabs getting many eros- https://i.imgur.com/Dr8iXYR.jpg
3.  Preview is not getting updated when changing the tabs recording in customizer  https://i.imgur.com/92JoBNg.jpg 
4. Changes are not getting saved after Save Settings.  https://i.imgur.com/MCRuDHN.jpg 

https://www.loom.com/share/4d5a33e93609416991901e417be8f892 

cc @mehulkaklotar 

**System Status-**  

Platform - respective branch.
Theme - Dev branch.
`
### WordPress Environment ###

WordPress address (URL): http://localhost:10049
Site address (URL): http://localhost:10049
WC Version: 4.5.2
REST API Version: ✔ 4.5.2
WC Blocks Version: ✔ 3.1.0
Action Scheduler Version: ✔ 3.1.6
WC Admin Version: ✔ 1.5.0
Log Directory Writable: ✔
WP Version: 5.5.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.16.0
PHP Version: 7.4.1
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 7.54.0
LibreSSL/2.6.5


### Active Plugins (4) ###

BuddyBoss Platform: by BuddyBoss – 1.5.1.1
Elementor Pro: by Elementor.com – 3.0.4
Elementor: by Elementor.com – 3.0.9
WooCommerce: by Automattic – 4.5.2

shohelbuddyboss commented:
@ankitguptaindia 

I didn't' face any of above issues, however, I fixed several issues. Please make a review by taking a pull.

ankitguptaindia commented:
> @ankitguptaindia
> 
> I didn't' face any of above issues, however, I fixed several issues. Please make a review by taking a pull.

@shohel 

The above issues reported have been fixed on my system as well. I am looking other cases now.
Thanks.

ankitguptaindia commented:
@shohelbuddyboss @mehulkaklotar 

**Status-** Feature has been implemented successfully and working fine. 

**Proof-** https://www.loom.com/share/a6c873cc6d16435e9757e2a6c33cfec9


FYI - If user want to reorder tabs coming from any 3rd party plugin then they have to use hook to make it orderable.

ankitguptaindia reviewed (approved):
@sourabhmatolia 

**Closing Remark-** 

- [x] Read PR description

- [x] Checked Description on related Issue

- [x] Pulled respective branch and checked module by pull stable branch (Checked the Before Fix & After Fix differences)

- [x] Tested negative test scenario

- [x] Updated Labels as per current status

**Final Update-**  Ready to Merge 🚀
